src/Nethermind/Nethermind.BeaconNode/AttestationProducer.cs

```diff
@@ -0,0 +1,126 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Nethermind.Core2;
+using Nethermind.Core2.Containers;
+using Nethermind.Core2.Crypto;
+using Nethermind.Core2.Types;
+
+namespace Nethermind.BeaconNode
+{
+    public class AttestationProducer
+    {
+        private readonly BeaconChainUtility _beaconChainUtility;
+        private readonly BeaconStateAccessor _beaconStateAccessor;
+        private readonly BeaconStateTransition _beaconStateTransition;
+        private readonly ICryptographyService _cryptographyService;
+        private readonly IForkChoice _forkChoice;
+        private readonly ILogger _logger;
+        private readonly IStore _store;
+
+        public AttestationProducer(ILogger<AttestationProducer> logger,
+            ICryptographyService cryptographyService,
+            BeaconChainUtility beaconChainUtility,
+            BeaconStateAccessor beaconStateAccessor,
+            BeaconStateTransition beaconStateTransition,
+            IForkChoice forkChoice,
+            IStore store)
+        {
+            _logger = logger;
+            _cryptographyService = cryptographyService;
+            _beaconChainUtility = beaconChainUtility;
+            _beaconStateAccessor = beaconStateAccessor;
+            _beaconStateTransition = beaconStateTransition;
+            _forkChoice = forkChoice;
+            _store = store;
+        }
+
+        public async Task<Attestation> NewAttestationAsync(BlsPublicKey validatorPublicKey,
+            bool proofOfCustodyBit, Slot slot, CommitteeIndex index,
+            CancellationToken cancellationToken)
+        {
+            Root head = await _store.GetHeadAsync().ConfigureAwait(false);
+            BeaconBlock headBlock = (await _store.GetSignedBlockAsync(head).ConfigureAwait(false)).Message;
+            BeaconState parentState = await _store.GetBlockStateAsync(head).ConfigureAwait(false);
+
+            // Clone state (will mutate) and process outstanding slots
+            BeaconState headState = BeaconState.Clone(parentState);
+            _beaconStateTransition.ProcessSlots(headState, slot);
+
+            // Set attestation_data.index = index where index is the index associated with the validator's committee.
+            ValidatorIndex? validatorIndex = FindValidatorIndexByPublicKey(headState, validatorPublicKey);
+
+            if (validatorIndex == null)
+            {
+                throw new Exception($"Can not find validator index for public key {validatorPublicKey}");
+            }
+
+            // TODO: May need a more efficient way to try and find the committee and position within the committee.
```

Comment:
I plan to look at things like this - efficient tree storage / retrieval / caching